File: 1475.Final-Prices-With-a-Special-Discount-in-a-Shop.js
/**
 * @param {number[]} prices
 * @return {number[]}
 */
var finalPrices = function (prices) {
  const result = [...prices];
  const stack = [];

  for (let i = 0; i < prices.length; i++) {
    const price = prices[i];

    while (stack.length && price <= prices[stack[stack.length - 1]]) {
      const idx = stack.pop();
      const diff = prices[idx] - price;
      result[idx] = diff;
    }

    stack.push(i);
  }

  return result;
};

File: python/1475.Final-Prices-With-a-Special-Discount-in-a-Shop.py
class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        stack = []
        result = prices[:]

        for i in range(len(prices)):
            while len(stack) and prices[i] <= prices[stack[-1]]:
                idx = stack[-1]
                result[idx] = prices[stack[-1]] - prices[i]
                stack.pop()
            stack.append(i)

        return result